Python/suggestions.c

```diff
@@ -0,0 +1,199 @@
+            char2p = string2;
+            D = x = i;
+        }
+        /* skip the lower triangle */
+        if (i <= half + 1) {
+            end = row + len2 + i - half - 2;
+        }
+        /* main */
+        while (scan_ptr <= end) {
+            size_t c3 = --D + (char1 != *(char2p++));
+            x++;
+            if (x > c3) {
+              x = c3;
+            }
+            D = *scan_ptr;
+            D++;
+            if (x > D)
+              x = D;
+            *(scan_ptr++) = x;
+        }
+        /* lower triangle sentinel */
+        if (i <= half) {
+            size_t c3 = --D + (char1 != *char2p);
+            x++;
+            if (x > c3) {
+              x = c3;
+            }
+            *scan_ptr = x;
+        }
+    }
+    i = *end;
+    PyMem_Free(row);
+    return i;
+}
+
+static inline PyObject*
+calculate_suggestions(PyObject* dir,
+                      PyObject* name)
+{
+    assert(!PyErr_Occurred());
+    assert(PyList_CheckExact(dir));
+
+    Py_ssize_t dir_size = PyList_GET_SIZE(dir);
+    if (dir_size >= MAX_GETATTR_PREDICT_ITEMS) {
+        return NULL;
+    }
+
+    Py_ssize_t suggestion_distance = PyUnicode_GetLength(name);
+    PyObject* suggestion = NULL;
+    for (int i = 0; i < dir_size; ++i) {
+        PyObject *item = PyList_GET_ITEM(dir, i);
+        const char *name_str = PyUnicode_AsUTF8(name);
+        if (name_str == NULL) {
+            PyErr_Clear();
+            continue;
+        }
+        Py_ssize_t current_distance = distance(PyUnicode_AsUTF8(name),
+                                               PyUnicode_AsUTF8(item));
+        if (current_distance > MAX_GETATTR_PREDICT_DIST){
+            continue;
+        }
+        if (!suggestion || current_distance < suggestion_distance) {
+            suggestion = item;
+            suggestion_distance = current_distance;
+        }
+    }
+    if (!suggestion) {
+        return NULL;
+    }
+    return PyUnicode_FromFormat("Did you mean: %U?", suggestion);
+}
+
+static PyObject*
+offer_suggestions_for_attribute_error(PyAttributeErrorObject* exc) {
+    PyObject* name = exc->name; // borrowed reference
+    PyObject* obj = exc->obj; // borrowed reference
+
+    // Abort if we don't have an attribute name or we have an invalid one
+    if (name == NULL || obj == NULL || !PyUnicode_CheckExact(name)) {
+        return NULL;
+    }
+
+    PyObject* dir = PyObject_Dir(obj);
+    if (dir == NULL) {
+        return NULL;
+    }
+
+    PyObject* suggestions = calculate_suggestions(dir, name);
+    Py_DECREF(dir);
+    return suggestions;
+}
+
+
+// Offer suggestions for a given exception. This function does not raise exceptions
+// and returns NULL if no exception was found.
+PyObject* _Py_Offer_Suggestions(PyObject* exception) {
+    assert(!PyErr_Occurred());
+    if (PyErr_GivenExceptionMatches(exception, PyExc_AttributeError)) {
+        return offer_suggestions_for_attribute_error((PyAttributeErrorObject*) exception);
```

Comment:
"This function does not raise exceptions"

Can you please add an assertion to ensure that?

PyUnicode_FromFormat() can fail with MemoryError.

IMO it's perfectly fine to raise an exception. The caller is free to decide how to handle it.